### PR TITLE
[REVIEW] Update to Arrow v1.0.1

### PR DESCRIPTION
**Pull request by trxcllnt**
This PR updates arrow-cpp and pyarrow to v1.0.1.

Depends on related integration PR: https://github.com/rapidsai/integration/pull/122

Closes https://github.com/rapidsai/cudf/issues/5835

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Review by ajschmidt8 — APPROVED** [review]

**Review by kkraus14 — APPROVED** [review]
CMake / Conda / Python LGTM

**Comment by kkraus14** [comment]
@rgsl888prabhu can you review the C++ as you most recently worked on a bunch of the `to_arrow` / `from_arrow` C++ code?

**Review by rgsl888prabhu — APPROVED** [review]
C++ looks good to me.

File: cpp/tests/interop/arrow_utils.hpp
Comment (cwharris) [inline cpp/tests/interop/arrow_utils.hpp:47]: Is it cool that we're using internal here? I don't know what Arrow's policies are. :man_shrugging:

**Review by cwharris — APPROVED** [review]

File: cpp/tests/interop/arrow_utils.hpp
Comment (trxcllnt) [reply cpp/tests/interop/arrow_utils.hpp:47]: It compiles and links, so I assume it's ok? This is also only part of the tests, but we could vendor this function if we need to.

File: cpp/CMakeLists.txt
Comment (trxcllnt) [inline cpp/CMakeLists.txt:432]: This change is to work around a CUDA 10.1 `nvcc` bug that only seems to happen on Ubuntu16.04. The Arrow change that exposed this bug is [here](https://github.com/apache/arrow/pull/7909/commits/239643e113346b36a3c73df983313fea3dcb1e04#diff-200337eb9b1aa7f7e176bdde8b3746d0R52-R56).

The fix in this commit splits `detail/concatenate.cuh` into two headers:
1. A `.hpp` file with the `detail::concatenate()` definitions (at `detail/concatenate.hpp`)
2. A `.cuh` file with the `detail::concatenate_masks()` definitions (at `detail/concatenate.cuh`)

The `detail/concatenate.cuh` header includes the new `detail/concatenate.hpp` so existing consumers of the `.cuh` header aren't affected.

Then `from_arrow.cu` is changed to include `detail/concatenate.hpp`. Then we rename `from_arrow.cu` -> `from_arrow.cpp`, so it's compiled by `gcc` instead of `nvcc`.

Here's the error we see in Ubuntu16.04/GCC7/CUDA 10.1:
```
include/arrow/array/array_nested.h:93:50: error: ‘arrow::BaseListArray’ is not a type
   friend void internal::SetListData<TYPE>(BaseListArray<TYPE>* self,
                                                  ^~~~~~~~~~~~~
include/arrow/array/array_nested.h: In instantiation of ‘class arrow::BaseListArray<arrow::ListType>’:
include/arrow/array/array_nested.h:103:63:   required from here
include/arrow/array/array_nested.h:93:13: error: template-id ‘SetListData<arrow::ListType>’ for ‘void arrow::internal::SetListData(int*, const std::shared_ptr<arrow::ArrayData>&, arrow::Type::type)’ does not match any template declaration
   friend void internal::SetListData<TYPE>(BaseListArray<TYPE>* self,
             ^~~~~~~~
include/arrow/array/array_nested.h:52:28: note: candidate is: template<class TYPE> void arrow::internal::SetListData(arrow::BaseListArray<TYPE>*, const std::shared_ptr<arrow::ArrayData>&, arrow::Type::type)
 template <typename TYPE>
                            ^          
include/arrow/array/array_nested.h: In instantiation of ‘class arrow::BaseListArray<arrow::LargeListType>’:
include/arrow/array/array_nested.h:147:68:   required from here
include/arrow/array/array_nested.h:93:13: error: template-id ‘SetListData<arrow::LargeListType>’ for ‘void arrow::internal::SetListData(int*, const std::shared_ptr<arrow::ArrayData>&, arrow::Type::type)’ does not match any template declaration
   friend void internal::SetListData<TYPE>(BaseListArray<TYPE>* self,
             ^~~~~~~~
include/arrow/array/array_nested.h:52:28: note: candidate is: template<class TYPE> void arrow::internal::SetListData(arrow::BaseListArray<TYPE>*, const std::shared_ptr<arrow::ArrayData>&, arrow::Type::type)
 template <typename TYPE>
                            ^          
```

CC: @rgsl888prabhu @harrism Is this an acceptable workaround for this issue?

**Comment by trxcllnt** [comment]
Not sure why we're seeing a bunch rpath-related issues from conda-build on the Ubuntu16.04 CPU builds.

Possibly related: https://github.com/conda/conda-build/issues/3588

```
ERROR :: get_rpaths_raw()=[] and patchelf=[''] disagree for /opt/conda/envs/rapids/conda-bld/libcudf_1599186085824/lib/libcudf.so ::  
WARNING (libcudf,lib/libcudf.so): Needed DSO lib/libstdc++.so.6 found in ['libstdcxx-ng']
WARNING (libcudf,lib/libcudf.so): .. but ['libstdcxx-ng'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
WARNING (libcudf,lib/libcudf.so): Needed DSO lib/libboost_filesystem.so.1.72.0 found in ['boost-cpp']
WARNING (libcudf,lib/libcudf.so): .. but ['boost-cpp'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
WARNING (libcudf,lib/libcudf.so): did not find - or even know where to look for: $RPATH/libpthread.so.0
```

**Comment by trxcllnt** [comment]
Changed this PR to WIP because @calebwin is going to contribute fixes for the parquet pytest failures.

File: cpp/CMakeLists.txt
Comment (rgsl888prabhu) [reply cpp/CMakeLists.txt:432]: This is looks good, I don't see any issue.

**Comment by kkraus14** [comment]
I think the current failures are due to us not specifying `boost-cpp` in the libcudf conda recipe as `arrow-cpp` used to have it but now seems to no longer have it

**Comment by trxcllnt** [comment]
It seems we're blocked by the `spdlog` conda version issue addressed by this PR: https://github.com/rapidsai/rmm/pull/542.

File: conda/recipes/libcudf/meta.yaml
Comment (kkraus14) [inline conda/recipes/libcudf/meta.yaml:28]: This should be changed to `>=3.17.0` before merge:
```suggestion
    - cmake >=3.17.0
```

File: conda/recipes/cudf_kafka/meta.yaml
Comment (kkraus14) [inline conda/recipes/cudf_kafka/meta.yaml:27]:
```suggestion
    - cmake >=3.17.0
```

File: conda/recipes/libcudf_kafka/meta.yaml
Comment (kkraus14) [inline conda/recipes/libcudf_kafka/meta.yaml:25]:
```suggestion
    - cmake >=3.17.0
```

File: ci/gpu/build.sh
Comment (kkraus14) [inline ci/gpu/build.sh:75]: This should be reverted before merge

**Review by kkraus14 — CHANGES_REQUESTED** [review]

File: ci/gpu/build.sh
Comment (kkraus14) [inline ci/gpu/build.sh:75]:
```suggestion
# conda remove -f rapids-build-env rapids-notebook-env
# conda install "your-pkg=1.0.0"
```

**Review by kkraus14 — APPROVED** [review]

**Pull request merged**
